packages/valory/skills/abstract_round_abci/tests/test_common.py

```diff
@@ -51,3 +63,367 @@ def test_random_selection() -> None:
+        # create a dummy `is_retries_exceeded` for `MagicMock`ed `randomness_api`
+        self.randomness_behaviour.context.randomness_api.is_retries_exceeded = (
+            lambda: retries_exceeded
+        )
+        gen = self.randomness_behaviour.async_act()
+
+        with mock.patch.object(
+            self.randomness_behaviour,
+            "failsafe_randomness",
+            dummy_generator(observation),
+        ), mock.patch.object(
+            self.randomness_behaviour,
+            "get_randomness_from_api",
+            dummy_generator(observation),
+        ), mock.patch.object(
+            self.randomness_behaviour,
+            "send_a2a_transaction",
+        ) as send_a2a_transaction_mocked, mock.patch.object(
+            self.randomness_behaviour,
+            "wait_until_round_end",
+        ) as wait_until_round_end_mocked, mock.patch.object(
+            self.randomness_behaviour,
+            "set_done",
+        ) as set_done_mocked, mock.patch.object(
+            self.randomness_behaviour,
+            "sleep",
+        ) as sleep_mocked:
+            next(gen)
+            if not failsafe_succeeds or failsafe_succeeds and observation is None:
+                last_iteration(gen)
+                return
+
+            # here, the observation is retrieved from either `failsafe_randomness` or `get_randomness_from_api`
+            # depending on the test's parametrization
+            if not observation:
+                last_iteration(gen)
+                sleep_mocked.assert_called_once_with(irrelevant_config["sleep_time"])
+                self.randomness_behaviour.context.randomness_api.increment_retries.assert_called_once()
+                return
+
+            last_iteration(gen)
+            send_a2a_transaction_mocked.assert_called_once()
+            wait_until_round_end_mocked.assert_called_once()
+            set_done_mocked.assert_called_once()
+
+    def test_clean_up(self) -> None:
+        """Test `clean_up`."""
+        self.randomness_behaviour.clean_up()
+        self.randomness_behaviour.context.randomness_api.reset_retries.assert_called_once()
+
+    def teardown(self) -> None:
+        """Teardown run after each test method."""
+        self.randomness_behaviour.context.randomness_api.increment_retries.reset_mock()
+
+
+class TestSelectKeeperBehaviour:
+    """Tests for `SelectKeeperBehaviour`."""
+
+    select_keeper_behaviour: DummySelectKeeperBehaviour
+
+    @classmethod
+    def setup_class(cls) -> None:
+        """Setup the test class."""
+        cls.select_keeper_behaviour = DummySelectKeeperBehaviour(
+            name="test",
+            skill_context=MagicMock(
+                params=BaseParams(
+                    name="test",
+                    skill_context=MagicMock(),
+                    service_id="test_id",
+                    consensus=dict(max_participants=1),
+                    **irrelevant_config,
+                )
+            ),
+        )
```

Comment:
put this in a base class, it is effectively a copy of this earlier code:
https://github.com/valory-xyz/open-autonomy/blob/1caae63e95274e19519fb415bce199d876a238fd/packages/valory/skills/abstract_round_abci/tests/test_common.py#L129-L145

replace `DummyRandomnessBehaviour` and `DummySelectKeeperBehaviour` in the `setup_class` with `cls.behaviour` and rename the class attribute to `behaviour`

Reply:
Refactored in fc2208c191d3b51a817d2821ab03082c71dd009b.